codec_default.go

```diff
@@ -0,0 +1,46 @@
+package avro
+
+import (
+	"unsafe"
+
+	"github.com/modern-go/reflect2"
+)
+
+func createDefaultDecoder(
+	cfg *frozenConfig,
+	schema Schema,
+	def any,
+	typ reflect2.Type,
+	w *Writer,
+	r *Reader,
+) ValDecoder {
+	defaultType := reflect2.TypeOf(def)
+	var defaultEncoder ValEncoder
+	// tmp workaround: codec_union failed to resolve name of struct{} typ
+	if def == nullDefault {
+		defaultEncoder = &nullCodec{}
+	} else {
+		defaultEncoder = encoderOfType(cfg, schema, defaultType)
+	}
+	if defaultType.LikePtr() {
+		defaultEncoder = &onePtrEncoder{defaultEncoder}
+	}
+	defaultEncoder.Encode(reflect2.PtrOf(def), w)
+
+	return &defaultDecoder{
+		defaultReader: r,
+		decoder:       decoderOfType(cfg, schema, typ),
+	}
```

Comment:
Regarding the **default decoder**,  I came up with an alternative that is less verbose and aligned with one of your first comments:

At Record decoder level, we can encode fields' default values and pipe the binary data to a new reader object `defaultReader` which is only used by fields with default values.

The draft is not optimized due to the encoding of defaults for every Record decoding operation, but I would like your thoughts on this approach first.

Reply:
Interesting approach. Other than the encoding issues, the big issue you have there is that the second call to the `Reader` will have already read past the data. It would be better to store the bytes and get a reader from `cfg.borrowReader` each time. I would also bring the writer inside the function, and take a field, like:
```
func createDefaultDecoder(cfg *frozenConfig, field *Field, typ reflect2.Type) ValDecoder {
```

I was hoping for a simpler way to deal with defaults, but I don't see one. From what I can see, you have landed on the simplest solution.